[MRG] FIX #10540 ElasticNet overwrites X even with copy_X=True

ymazari opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes #10540. Closes #10568.
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
if copy_X=True:
- Do not overwrite X
- Ensure copying happens only once: either during check_input or _pre_fit

#### Any other comments?
- Added unit tests for verification.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

jnothman commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -669,6 +669,27 @@ def test_check_input_false():
+
+    original_X = X.copy()
+    enet = ElasticNet(copy_X=True)
+    enet.fit(X, y, check_input=False)
```
I'd use a loop or pytest.mark.parametrize to vary check_input, to clearly show to the reader what invariant you are trying to uphold.

ymazari replied:
Good suggestion. Added pytest.mark.parametrize.

jnothman reviewed (commented):
Otherwise LGTM, thanks

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -700,19 +700,21 @@ def fit(self, X, y, check_input=True):
         X, y, X_offset, y_offset, X_scale, precompute, Xy = \
             _pre_fit(X, y, None, self.precompute, self.normalize,
-                     self.fit_intercept, copy=False)
+                     self.fit_intercept, copy=not X_copied)
```
logic is not correct. If you pass check_input=False in fit and copy_X=False in init this code will still make a copy.

ymazari replied:
This bug is trickier than what it initially looked like ;)
I fixed the logic to handle that case.
Thanks a lot for the useful feedback!!

ymazari commented:
Aaaah, fix one thing and break 20 others. I am checking/fixing the failures.

ymazari commented:
@agramfort I made the changes based on your feedback, could you please review it?
@jnothman I switched this PR back to "MRG".

agramfort commented:
LGTM

please update what's new with an entry in the bug fix section.

thanks

jnothman reviewed (commented):
Nice.

lesteve commented:
@ymazari can you add an entry in doc/whats_new/v0.20.rst in the "bug fixes" section as requested by @agramfort? This one can be merged when you do this.

ymazari committed:
[10540] ElasticNet overwrites X even with copy_X=True
- Set copy=True if copy_X=True
- Ensure copying happens only once
- Fix logic with when both copy_X and check_input are False
- Add unit tests
- Update v0.20.rst

ymazari commented:
@lesteve, @agramfort: I updated   v0.20.rst.
This is now ready to be merged :)

agramfort commented on doc/whats_new/v0.20.rst:
```diff
@@ -234,6 +234,10 @@ Classifiers and regressors
   near infinite. :issue:`10536` by :user:`Jonathan Ohayon <Johayon>`.
 
+- Fixed a bug in :class:`linear_model.ElasticNet` which caused the input
+  to be overridden even when using parameter ``copy_X=True``.
```
and ``check_input=False``

glemaitre commented:
by the way it should close #10568 I think

lesteve committed:
tweaks

sklearn-lgtm commented:
This pull request introduces 37 alerts and fixes 18 - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-e8d9c650c251b9e5b2e3944fa52615437efd4e7f)

**new alerts:**

* 37 for Explicit export is not defined

**fixed alerts:**

* 6 for Mismatch between signature and use of an overridden method
* 3 for Missing call to __init__ during object initialization
* 3 for Potentially uninitialized local variable
* 2 for Comparison using is when operands support __eq__
* 2 for Wrong number of arguments for format
* 1 for Conflicting attributes in base classes
* 1 for Mismatch between signature and use of an overriding method

---

*Comment posted by [lgtm.com](https://lgtm.com)*
